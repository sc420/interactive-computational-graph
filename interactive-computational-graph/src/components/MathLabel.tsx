import { Box, Link, Typography } from "@mui/material";
import { type FunctionComponent } from "react";
import type MathLabelPartType from "../features/MathLabelPartType";
import Katex from "../latex/Katex";

interface MathLabelProps {
  parts: MathLabelPartType[];
  onClickLatexLink: (nodeId: string) => void;
}

const MathLabel: FunctionComponent<MathLabelProps> = ({
  parts,
  onClickLatexLink,
}) => {
  return (
    <Box>
      {parts.map((part) => {
        switch (part.type) {
          case "text": {
            return (
              <Typography key={part.id} display="inline" variant="body2">
                {part.text}
              </Typography>
            );
          }
          case "latex": {
            return <Katex key={part.id} latex={part.latex} />;
          }
          case "latexLink": {
            return (
              <Link
                key={part.id}
                href="#"
                onClick={() => {
                  onClickLatexLink(part.href);
                }}
              >
                <Katex latex={part.latex} />
              </Link>
            );
          }
          default: {
            throw new Error("Unknown description part type");
          }
        }
      })}
    </Box>
  );
};

export default MathLabel;
