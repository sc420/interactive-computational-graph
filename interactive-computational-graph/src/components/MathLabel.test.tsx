import { fireEvent, render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import type MathLabelPartType from "../features/MathLabelPartType";
import MathLabel from "./MathLabel";

jest.mock("../latex/Katex");

test("should render correctly with mixed types", () => {
  const parts: MathLabelPartType[] = [
    {
      type: "text",
      id: "id1",
      text: "To calculate ",
    },
    {
      type: "latex",
      id: "id2",
      latex: "x",
    },
    {
      type: "text",
      id: "id3",
      text: ", we need to calculate ",
    },
    {
      type: "latexLink",
      id: "id4",
      latex: "\\frac{a}{b}",
      href: "123",
    },
  ];
  const handleClickLatexLink = jest.fn();
  const tree = renderer
    .create(<MathLabel parts={parts} onClickLatexLink={handleClickLatexLink} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("should trigger event when clicking the latex link", () => {
  const parts: MathLabelPartType[] = [
    {
      type: "latexLink",
      id: "link",
      latex: "\\frac{a}{b}",
      href: "123",
    },
  ];
  const handleClickLatexLink = jest.fn();
  render(<MathLabel parts={parts} onClickLatexLink={handleClickLatexLink} />);

  const latexLink = screen.getByRole("link");
  fireEvent.click(latexLink);
  expect(handleClickLatexLink).toHaveBeenCalledWith("123");
});
