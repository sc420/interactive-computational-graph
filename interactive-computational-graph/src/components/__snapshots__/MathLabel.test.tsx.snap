// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly with mixed types 1`] = `
<div
  className="MuiBox-root css-0"
>
  <p
    className="MuiTypography-root MuiTypography-body2 css-141vi1f-MuiTypography-root"
  >
    To calculate 
  </p>
  <span>
    x
  </span>
  <p
    className="MuiTypography-root MuiTypography-body2 css-141vi1f-MuiTypography-root"
  >
    , we need to calculate 
  </p>
  <a
    className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1bhi0za-MuiTypography-root-MuiLink-root"
    href="#"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
  >
    <span>
      \\frac{a}{b}
    </span>
  </a>
</div>
`;
